ClientMapStoreTest Timeout

```
02:37:41 Running com.hazelcast.client.map.ClientMapStoreTest
04:02:32 Build timed out (after 360 minutes). Marking the build as failed.
```

Main reason of aboreted failures is this timeout problem.

Here is the trend for this build:

<img width="332" alt="screen shot 2015-12-07 at 12 27 37" src="https://cloud.githubusercontent.com/assets/6005622/11624642/dd02a960-9cdd-11e5-84ba-b9072aa7d67d.png">

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-nightly/686/console

We seem to have this issue back in the current master
```
21:13:28 Running com.hazelcast.client.map.ClientMapStoreTest
02:37:28 Build timed out (after 480 minutes). Marking the build as failed.
02:37:28 
02:37:28 Results :
02:37:28 
02:37:28 Tests run: 1305, Failures: 0, Errors: 0, Skipped: 14
02:37:28 
02:37:28 [ERROR] There was a timeout or other error in the fork
02:37:28 [WARNING] Failed to notify spy hudson.maven.Maven3Builder$JenkinsEventSpy: java.util.concurrent.ExecutionException: Invalid object ID 6 iota=17
02:37:28 [WARNING] Failed to notify spy hudson.maven.Maven3Builder$JenkinsEventSpy: java.util.concurrent.ExecutionException: Invalid object ID 6 iota=17
02:37:28 [WARNING] Failed to notify spy hudson.maven.Maven3Builder$JenkinsEventSpy: java.util.concurrent.ExecutionException: Invalid object ID 9 iota=17
```
https://hazelcast-l337.ci.cloudbees.com/view/Sonar/job/Hazelcast-3.x-sonar/1394/consoleFull

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapKeyLoader.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapKeyLoader.java
@@ -214,12 +214,43 @@ public class MapKeyLoader {
                 public void run() {
                     // checks if loading has finished and triggers loading in case SENDER died and SENDER_BACKUP took over.
                     Operation op = new TriggerLoadIfNeededOperation(mapName);
-                    opService.<Boolean>invokeOnPartition(SERVICE_NAME, op, mapNamePartition);
+                    opService.<Boolean>invokeOnPartition(SERVICE_NAME, op, mapNamePartition)
+                            // required since loading may be triggerd after migration
+                            // and in this case the callback is the only way to get to know if the key load finished or not.
+                            .andThen(loadingFinishedCallback());
                 }
             });
         }
 
         return keyLoadFinished;
+    }
+
+    private ExecutionCallback<Boolean> loadingFinishedCallback() {
+        return new ExecutionCallback<Boolean>() {
+            @Override
+            public void onResponse(Boolean loadingFinished) {
+                if (loadingFinished) {
+                    updateLocalKeyLoadStatus(null);
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+                updateLocalKeyLoadStatus(t);
+            }
+        };
+    }
+
+    private void updateLocalKeyLoadStatus(Throwable t) {
+        Operation op = new KeyLoadStatusOperation(mapName, t);
+        // This updates the local record store on the partition thread.
+        // If invoked by the SENDER_BACKUP however it's the replica index has to be set to 1, otherwise
+        // it will be a remote call to the SENDER who is the owner of the given partitionId.
+        if (hasBackup && role.is(Role.SENDER_BACKUP)) {
+            opService.createInvocationBuilder(SERVICE_NAME, op, partitionId).setReplicaIndex(1).invoke();
+        } else {
+            opService.createInvocationBuilder(SERVICE_NAME, op, partitionId).invoke();
+        }
     }
 
     public Future<?> startLoading(MapStoreContext mapStoreContext, boolean replaceExistingValues) {

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/operation/TriggerLoadIfNeededOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/operation/TriggerLoadIfNeededOperation.java
@@ -22,6 +22,8 @@ import com.hazelcast.spi.ReadonlyOperation;
 
 public class TriggerLoadIfNeededOperation extends MapOperation implements PartitionAwareOperation, ReadonlyOperation {
 
+    private Boolean isLoaded;
+
     public TriggerLoadIfNeededOperation() {
     }
 
@@ -31,11 +33,17 @@ public class TriggerLoadIfNeededOperation extends MapOperation implements Partit
 
     @Override
     public void run() {
+        isLoaded = recordStore.isLoaded();
         recordStore.maybeDoInitialLoad();
     }
 
+    @Override
+    public Object getResponse() {
+        return isLoaded;
+    }
+
     public boolean returnsResponse() {
-        return false;
+        return true;
     }
 
     @Override

